src/sentry/migrations/0269_delete_legacy_alert_rule_action_records.py

```diff
@@ -0,0 +1,97 @@
+# Generated by Django 2.2.24 on 2022-01-27 00:00
+
+import pickle
+
+from django.db import migrations
+
+from sentry.new_migrations.migrations import CheckedMigration
+from sentry.utils.query import RangeQuerySetWrapperWithProgressBar
+from sentry.utils.strings import decompress
+
+
+def delete_metric_alert_trigger_actions_with_alert_rule_actions(apps, schema_editor):
+    """
+    Delete the rows in the AlertRuleTriggerAction table that use an Alert Rule Action SentryApp.
+    """
+    AlertRuleTriggerAction = apps.get_model("sentry", "AlertRuleTriggerAction")
+    SentryAppComponent = apps.get_model("sentry", "SentryAppComponent")
+
+    for trigger_action in AlertRuleTriggerAction.objects.filter(
+        sentry_app_id__in=SentryAppComponent.objects.filter(type="alert-rule-action").values_list(
+            "sentry_app_id", flat=True
+        ),
+        sentry_app_config__isnull=False,
+    ):
+        trigger_action.delete()
+
+
+def delete_issue_alerts_with_alert_rule_actions(apps, schema_editor):
+    """
+    Delete the rows in the Rule table that use an Alert Rule Action SentryApp.
+    """
+    SentryAppComponent = apps.get_model("sentry", "SentryAppComponent")
+    SentryAppInstallation = apps.get_model("sentry", "SentryAppInstallation")
+    Project = apps.get_model("sentry", "Project")
+    Rule = apps.get("sentry", "Rule")
+
+    # First find all projects by orgs that have the sentryapps with Alert Rule Actions installed
+    # Filter sentry_rule table by the affected projects.
+    # Unpickle each rule to determine if they have an Alert Rule Action and then delete
+    for rule in RangeQuerySetWrapperWithProgressBar(
```

Comment:
I think making this a single query would look like 
```py
query = Rule.objects.filter(
    project__organization__sentry_app_installations__sentry_app__components__type="alert-rule-action"
)
for rule in query:
    ...
```
or 
```py
for rule in RangeQuerySetWrapperWithProgressBar(Rule.objects.all()):
    installs = rule.project.organization.sentry_apps.sentry_app_installations.all()
    if any([install.sentry_app.components.type == "alert-rule-action" for install in installs]):
        ...
```

Reply:
@mgaeta The first query is just doing inner joins. It's not as good as the original query for reducing the number of rows to iterate over. Outputs:
```
SELECT sentry_rule.id, sentry_rule.project_id, sentry_rule.environment_id, sentry_rule.label, sentry_rule.data, sentry_rule.status, sentry_rule.owner_id, sentry_rule.date_added
        FROM sentry_rule
        INNER JOIN sentry_project ON
    (sentry_rule.project_id = sentry_project.id)
    INNER JOIN sentry_organization ON
    (sentry_project.organization_id = sentry_organization.id)
    INNER JOIN sentry_sentryappinstallation ON
    (sentry_organization.id = sentry_sentryappinstallation.organization_id) INNER JOIN sentry_sentryapp
    ON (sentry_sentryappinstallation.sentry_app_id = sentry_sentryapp.id) INNER JOIN sentry_sentryappcomponent
    ON (sentry_sentryapp.id = sentry_sentryappcomponent.sentry_app_id) WHERE sentry_sentryappcomponent.type = 'alert-rule-action'
```
1563 rows

Original Query: 182 rows

The second option looks like it'll loop over all the records in `sentry_rule` which 9623311 rows. And also FK relations are the other way around. Rule -> Project -> Organization <- Sentry_App_Installation <- Sentry_App <- Sentry_App_Component